src/domain/notifications/v2/notifications.repository.ts

```diff
@@ -88,19 +92,188 @@ export class NotificationsRepositoryV2 implements INotificationsRepositoryV2 {
     return isNotFound && isUnregistered;
   }
 
-  async upsertSubscriptions(args: {
+  public async upsertSubscriptions(args: {
     authPayload: AuthPayload;
     upsertSubscriptionsDto: UpsertSubscriptionsDto;
   }): Promise<{
     deviceUuid: UUID;
   }> {
-    return this.notificationsDatasource.upsertSubscriptions({
-      signerAddress: args.authPayload.signer_address,
-      upsertSubscriptionsDto: args.upsertSubscriptionsDto,
+    let device;
+    await this.postgresDatabaseService.transaction(async () => {
+      device = await this.upsertDevice(args);
+      await this.deletePreviousSubscriptions({
+        deviceId: device.id,
+        signerAddress: args.authPayload.signer_address,
+        upsertSubscriptionsDto: args.upsertSubscriptionsDto,
+      });
+
+      const subscriptions = await this.upsertSubscription({
+        ...args,
+        deviceId: device.id,
+      });
+
+      await this.insertSubscriptionNotificationTypes({
+        subscriptions,
+        upsertSubscriptionsDto: args.upsertSubscriptionsDto,
+      });
+    });
+
+    return { deviceUuid: device!.device_uuid };
```

Comment:
This is a question, the implementation is fine. Do you know if it's possible to return a value from the transaction callback? I'm intrigued as to _why_ TypeScript isn't inferring device correctly. Even if we initially type `device`, it's not picking up the assignment.

Reply:
Done:
https://github.com/safe-global/safe-client-gateway/pull/2063/commits/33b92e8baea99e8691b01ddc081e13eb726cf18f